framework/src/mesh/SpiralAnnularMesh.C

```diff
@@ -0,0 +1,306 @@
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "SpiralAnnularMesh.h"
+
+#include "libmesh/face_quad4.h"
+#include "libmesh/face_tri3.h"
+
+registerMooseObject("MooseApp", SpiralAnnularMesh);
+
+template <>
+InputParameters
+validParams<SpiralAnnularMesh>()
+{
+  InputParameters params = validParams<MooseMesh>();
+  params.addRequiredRangeCheckedParam<Real>(
+      "inner_radius", "inner_radius>0.", "The size of the inner circle.");
+  params.addRequiredRangeCheckedParam<Real>("outer_radius",
+                                            "outer_radius>0.",
+                                            "The size of the outer circle."
+                                            " Logically, it has to be greater than inner_radius");
+  params.addRequiredRangeCheckedParam<unsigned int>(
+      "nodes_per_ring", "nodes_per_ring>0", "Number of nodes on each ring.");
+  params.addParam<bool>(
+      "my_second_order", false, "Generate mesh of TRI6 elements instead of TRI3 elements.");
+  params.addRequiredRangeCheckedParam<unsigned int>(
+      "num_rings", "num_rings>0", "The number of rings.");
+  params.addParam<boundary_id_type>(
+      "cylinder_bid", 1, "The boundary id to use for the cylinder (inner circle)");
+  params.addParam<boundary_id_type>(
+      "exterior_bid", 2, "The boundary id to use for the exterior (outer circle)");
+  params.addParam<Real>("initial_delta_r",
+                        "Width of the initial layer of elements around the cylinder."
+                        "This number should be approximately"
+                        " 2 * pi * inner_radius / nodes_per_ring to ensure that the"
+                        " initial layer of elements is almost equilateral");
+  params.addClassDescription("Creates an annual mesh based on TRI3 elements"
+                             " (it can also be TRI6 elements) on several rings.");
+
+  return params;
+}
+
+SpiralAnnularMesh::SpiralAnnularMesh(const InputParameters & parameters)
+  : MooseMesh(parameters),
+    _inner_radius(getParam<Real>("inner_radius")),
+    _outer_radius(getParam<Real>("outer_radius")),
+    _radial_bias(1.0),
+    _nodes_per_ring(getParam<unsigned int>("nodes_per_ring")),
+    _second_order(getParam<bool>("my_second_order")),
+    _num_rings(getParam<unsigned int>("num_rings")),
+    _cylinder_bid(getParam<boundary_id_type>("cylinder_bid")),
+    _exterior_bid(getParam<boundary_id_type>("exterior_bid")),
+    _initial_delta_r(2 * libMesh::pi * _inner_radius / _nodes_per_ring)
+{
+
+  // catch likely user errors
+  if (_outer_radius <= _inner_radius)
+    mooseError("SpiralAnnularMesh: outer_radius must be greater than inner_radius");
+  if (_num_rings == 0)
+    mooseError("Cannot create mesh with _num_rings == 0");
```

Comment:
@leorachapuis I also have not tested this code with an odd-numbered `_nodes_per_ring` parameter. I don't see any reason why it wouldn't work, but it would still be good to test...

Reply:
The same comment applies to the `_num_rings` parameter: I believe I have only tested with even numbers of rings, but an odd number should also be possible.